trie-db/src/triedbmut.rs

```diff
@@ -1804,152 +2034,120 @@ where
+				)));
+				Changeset {
+					root: ChangesetNodeRef::Existing(ExistingChangesetNode {
+						hash,
+						prefix: Default::default(),
+						location,
+					}),
+					removed,
+				}
 			},
 		}
 	}
 
 	/// Cache the given `encoded` node.
-	fn cache_node(&mut self, hash: TrieHash<L>, encoded: &[u8], full_key: Option<NibbleVec>) {
-		// If we have a cache, cache our node directly.
+	fn cache_node(&mut self, hash: TrieHash<L>) {
+		// Mark the node as new so that it is removed from the shared cache.
 		if let Some(cache) = self.cache.as_mut() {
-			let node = cache.get_or_insert_node(hash, &mut || {
-				Ok(L::Codec::decode(&encoded)
-					.ok()
-					.and_then(|n| n.to_owned_node::<L>().ok())
-					.expect("Just encoded the node, so it should decode without any errors; qed"))
-			});
-
-			// `node` should always be `OK`, but let's play it safe.
-			let node = if let Ok(node) = node { node } else { return };
-
-			let mut values_to_cache = Vec::new();
-
-			// If the given node has data attached, the `full_key` is the full key to this node.
-			if let Some(full_key) = full_key {
-				node.data().and_then(|v| node.data_hash().map(|h| (&full_key, v, h))).map(
-					|(k, v, h)| {
-						values_to_cache.push((k.inner().to_vec(), (v.clone(), h).into()));
-					},
-				);
-
-				fn cache_child_values<L: TrieLayout>(
-					node: &NodeOwned<TrieHash<L>>,
-					values_to_cache: &mut Vec<(Vec<u8>, CachedValue<TrieHash<L>>)>,
-					full_key: NibbleVec,
-				) {
-					node.child_iter().flat_map(|(n, c)| c.as_inline().map(|c| (n, c))).for_each(
-						|(n, c)| {
-							let mut key = full_key.clone();
-							n.map(|n| key.push(n));
-							c.partial_key().map(|p| key.append(p));
-
-							if let Some((hash, data)) =
-								c.data().and_then(|d| c.data_hash().map(|h| (h, d)))
-							{
-								values_to_cache
-									.push((key.inner().to_vec(), (data.clone(), hash).into()));
-							}
-
-							cache_child_values::<L>(c, values_to_cache, key);
-						},
-					);
-				}
-
-				// Also cache values of inline nodes.
-				cache_child_values::<L>(&node, &mut values_to_cache, full_key.clone());
-			}
-
-			values_to_cache.into_iter().for_each(|(k, v)| cache.cache_value_for_key(&k, v));
+			cache.insert_new_node(&hash);
 		}
 	}
 
 	/// Cache the given `value`.
 	///
 	/// `hash` is the hash of `value`.
-	fn cache_value(&mut self, full_key: &[u8], value: impl Into<Bytes>, hash: TrieHash<L>) {
-		if let Some(cache) = self.cache.as_mut() {
-			let value = value.into();
-
-			// `get_or_insert` should always return `Ok`, but be safe.
-			let value = if let Ok(value) = cache
-				.get_or_insert_node(hash, &mut || Ok(NodeOwned::Value(value.clone(), hash)))
-				.map(|n| n.data().cloned())
-			{
-				value
-			} else {
-				None
-			};
-
-			if let Some(value) = value {
-				cache.cache_value_for_key(full_key, (value, hash).into())
-			}
-		}
-	}
+	fn cache_value(&mut self, _full_key: &[u8], _value: impl Into<Bytes>, _hash: TrieHash<L>) {}
```

Comment:
Generally arkpar/new-trie seems to have disable/remove some cache usage with triedbmut, is it something that should be restore later?

Reply:
Basically we can't cache a node before we know its location info. Otherwise if we add a node with no location info, it may be later retrieved from the cache and the database will fail to locate children.